clusterloader2/pkg/state/resource_state.go

```diff
@@ -0,0 +1,73 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package state
+
+import (
+	"fmt"
+	"strconv"
+	"sync"
+)
+
+// ResourceIdentifier is a unique identifier for resource
+type ResourceIdentifier struct {
+	ObjectKind string
+	ApiGroup   string
+}
+
+// resourcesVersionsState represents most recent resources versions for a given object types.
+// These versions are available only for resource types of objects that were create/updated
+// by the clusterloader.
+type resourcesVersionsState struct {
+	lock              sync.RWMutex
+	resourcesVersions map[ResourceIdentifier]uint64
+}
+
+// newResourcesVersionState creates new resources versions state.
+func newResourcesVersionsState() *resourcesVersionsState {
+	return &resourcesVersionsState{
+		resourcesVersions: make(map[ResourceIdentifier]uint64),
+	}
+}
+
+// Get returns state of current resource version.
+func (rs *resourcesVersionsState) Get(identifier ResourceIdentifier) (string, bool) {
+	rs.lock.RLock()
+	defer rs.lock.RUnlock()
+	versionUInt64, exists := rs.resourcesVersions[identifier]
+	if !exists {
+		return "", false
```

Comment:
Maybe that's fine, but I'm thinking about two other possible apis:
- return just string (and in this case, default to "0") [though it has disadvantages]
- returning (string, error) and returning notFound here (that makes it more extendable in the future if needed).

Reply:
The reasons I would prefer to keep current api:
- It similar to what golang map has.
- `namespaceState` has the same api. IMHO having two similar structs with two different apis sounds bad.

We can change both apis. The question here is whether implementing `notFound` error brings anything new.

Reply:
It's more clean in my opinion, but I'm not going to insist on that. My arguments are comparably weak to yours, to let's not change anything :-)